src/backend/cdb/dispatcher/cdbdisp_async.c

```diff
@@ -680,11 +685,37 @@ checkDispatchResult(CdbDispatcherState *ds, int timeout_sec)
 			if (timeout_sec >= 0 && diff_us >= timeout_sec * 1000000L)
 				break;
 		}
+		else if (should_cancel)
+		{
+			/* Make QEs come to their senses. */
+			signalQEs(pParms);
+
+			for (i = 0; i < db_count; i++)
+			{
+				dispatchResult = pParms->dispatchResultPtrArray[i];
+
+				/*
+				 * If we're cancelling the transaction due to an OOM, there
+				 * might not be enough memory to discard the result properly.
+				 * Try to free the async results first and pray.
+				 */
+				pqClearAsyncResult(dispatchResult->segdbDesc->conn);
+				cdbconn_discardResults(dispatchResult->segdbDesc, 20);
```

Comment:
We can still get error recursion if we are unlucky enough. For example, if the tuple is wide enough and we have several parallel queries. How acceptable is this for us? Maybe we should add the ability to read the result from segments without parsing and memory allocation?

Reply:
Looks like this has been fixed